### PR TITLE
feature: Move TileDB backend engine for xarray from TileDB-xarray to this repo

### PR DESCRIPTION
**Pull request by jp-dark**

### CODE REVIEW

File: tests/engines/xarray_engine/conftest.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            tiledb.Dim(name="x", domain=(1, 4), tile=4, dtype=np.int32),
+        ),
+        sparse=False,
+        attrs=[
+            tiledb.Attr(name="count", dtype=np.int32),
+            tiledb.Attr(name="pressure", dtype=np.float64),
+        ],
+    )
+    tiledb.DenseArray.create(array_uri, schema)
+    with tiledb.DenseArray(array_uri, mode="w") as array:
```
Comment (gsakkis) [inline tests/engines/xarray_engine/conftest.py]: Using `DenseArray` directly is discouraged; better use `tiledb.Array.create` and `with tiledb.open` instead (same comment for all `DenseArray` occurences).

File: tests/engines/xarray_engine/conftest.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        array.meta["global_2"] = "value2"
+        array.meta["__tiledb_attr.float_data.long_name"] = "example float data"
+        array.meta["__tiledb_attr.int_data.long_name"] = "example int data"
+    return array_uri, expected
+
+
+@pytest.fixture
+def create_tiledb_datetime_example(tmpdir):
+    _data = np.linspace(-1.0, 20.0, num=16, endpoint=True, dtype=np.float64)
+    _date = np.arange(np.datetime64("2000-01-01"), np.datetime64("2000-01-17"))
```
Comment (gsakkis) [inline tests/engines/xarray_engine/conftest.py]: Are the underscores needed for `_data` and `_date`?

File: tests/engines/xarray_engine/test_plugin.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            slice(5, 0, -2),
+            slice(1, 1),
+            np.array([0, 5, 3, 2, 1]),
+            np.array([-1, -5, -3, -2, -15]),
+            np.array([0, 0, 2, 2, 1, 3]),
+        ],
+    )
+    def test_indexing_array_1D(self, simple_data_arrays, index):
+        (tiledb_data_array, xarray_data_array) = simple_data_arrays
+        xr.testing.assert_allclose(tiledb_data_array, xarray_data_array)
```
Comment (gsakkis) [inline tests/engines/xarray_engine/test_plugin.py]: Where is the parametrized `index` used? Likewise for `test_indexing_array_2D`

File: tiledb/cf/engines/xarray_engine.py [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+        xarray_indices = indexer.tuple
+        if len(xarray_indices) != len(self._index_converters):
+            raise ValueError(
+                f"key of length {len(xarray_indices)} cannot be used for a TileDB array"
+                f" of length {len(self._index_converters)}"
+            )
+        shape = tuple(
+            len(range(converter.size)[index] if isinstance(index, slice) else index)
+            for index, converter in zip(xarray_indices, self._index_converters)
+            if not isinstance(index, integer_types)
```
Comment (gsakkis) [inline tiledb/cf/engines/xarray_engine.py:266]: Why exclude integer indexes here?

Comment (jp-dark) [reply tiledb/cf/engines/xarray_engine.py:266]: In xarray, if a dimension has an integer look-up, then it collapses the dimension on that value. For example,

```
x = xr.DataArray(np.arange(16).reshape(4,4))
print(x[0, :])
```
will yield
```
<xarray.DataArray (dim_1: 4)>
array([0, 1, 2, 3])
Dimensions without coordinates: dim_1
```

**Review by gsakkis — APPROVED** [review]

**Pull request merged**